### PR TITLE
JoepegAuctionHouse

### PR DESCRIPTION
**Pull request by 0xnivek**
This PR proposes a `JoepegAuctionHouse.sol` which supports running **English and Dutch Auctions on ERC721 tokens**. I'm putting up this PR to get feedback while I work on:

- [x] Tests
- [x] Deploy scripts

To see examples of how auctions are run on other marketplaces, see the **References** section of https://www.notion.so/traderjoexyz/Create-auction-bid-flow-dd5bdb8cacb34874852eb8c36d4a0275

### CODE REVIEW

File: contracts/JoepegAuctionHouse.sol [hunk 1]
```diff
@@ -0,0 +1,960 @@
...
+        address currency;
+        address lastBidder;
+        uint256 lastBidPrice;
+        uint256 endTime;
+        uint256 startPrice;
+    }
+
+    uint256 public immutable PERCENTAGE_PRECISION = 10000;
+
+    address public WAVAX;
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:72]: this sounds good to me but had quick question — I thought upgradeable contracts couldn't have constructors. is this safe to do?

File: contracts/JoepegAuctionHouse.sol [hunk 2]
```diff
@@ -0,0 +1,960 @@
...
+    event RoyaltyPayment(
+        address indexed collection,
+        uint256 indexed tokenId,
+        address indexed royaltyRecipient,
+        address currency,
+        uint256 amount
+    );
+
+    modifier isSupportedCurrency(address _currency) {
+        if (!currencyManager.isCurrencyWhitelisted(_currency)) {
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:189]: yeah `WAVAX` is allowlisted inside `currencyManager` already. I think I prefer to keep logic like this in case there is  a scenario where we want to turn it off for some reason and also just to keep the source of truth to be `CurrencyManager`

File: contracts/JoepegAuctionHouse.sol [hunk 3]
```diff
@@ -0,0 +1,960 @@
...
+
+    /// @notice Starts an English Auction for an ERC721 token
+    /// @dev Note this requires the auction house to hold the ERC721 token in escrow
+    /// @param _collection address of ERC721 token
+    /// @param _tokenId token id of ERC721 token
+    /// @param _currency address of currency to sell ERC721 token for
+    /// @param _duration number of seconds for English Auction to run
+    /// @param _startPrice minimum starting bid price
+    function startEnglishAuction(
+        address _collection,
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:236]: I kept it as `address` for some callsites b/c I was using it as `address` multiple times in the function body but only needed to convert to `IERC721` once so thought its cleaner this way.
slightly tangential to this, do you think I should add a check in the beginning of the function to ensure that the collection conforms to ERC721 interface (i.e. isn't ERC1155 token)? I believe since I use `IERC721.safeTransferFrom` it will revert if it isnt ERC721 but that happens at the end of the function

File: contracts/JoepegAuctionHouse.sol [hunk 4]
```diff
@@ -0,0 +1,960 @@
...
+        uint256 _tokenId,
+        address _currency,
+        uint256 _duration,
+        uint256 _startPrice
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:240]: yes I think so but we probably don't want to allow bids of 0 so I added a check for `_bidAmount != 0` in the `_placeEnglishAuctionBid` function

File: contracts/JoepegAuctionHouse.sol [hunk 5]
```diff
@@ -0,0 +1,960 @@
...
+        if (
+            msg.sender != auction.creator && block.timestamp < auction.endTime
+        ) {
+            revert JoepegAuctionHouse__EnglishAuctionOnlyCreatorCanSettleBeforeEndTime();
+        }
+
+        _clearEnglishAuction(_collection, _tokenId);
+
+        // Settle auction using latest bid
+        if (auction.currency == WAVAX) {
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:360]: hm thats a good question. maybe we could enforce caller is EOA? @LouisMeMyself do you have any thoughts here?

**Comment by 0xnivek** [comment]
thanks for all the thorough feedback @LouisMeMyself @simsk87 @Mathieu-Be ! 🙏  I addressed your feedback + left some replies

File: contracts/JoepegAuctionHouse.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:360]: another idea just in case of emergency is adding an `onlyOwner` function that allows owner to end an auction (i.e. return NFT to seller and bid to bidder)

File: contracts/JoepegAuctionHouse.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoepegAuctionHouse.sol:72]: Upgradeable contracts can have constructor, it's just that the values that aren't stored within the bytecode (any values that aren't immutable or constants), will not have the expected value because it will load the proxy's storage slot instead of the implementation ones.
That also means that it can't be changed. For example, if you use clone pattern, all cloned contract would have the same values for the immutable and constant because the constructor would not be run again.

File: contracts/JoepegAuctionHouse.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoepegAuctionHouse.sol:189]: Currently the specific AVAX/WAVAX functions doesn't check the currencyManager, so there is no way to turn wavax/avax off

File: contracts/JoepegAuctionHouse.sol [hunk 6]
```diff
@@ -90,7 +93,7 @@ contract JoepegAuctionHouse is
     /// - Number of seconds before an English Auction ends where any new
     ///   bid will extend the auction's end time
     /// - Number of seconds to extend an English Auction's end time by
-    uint256 public englishAuctionRefreshTime;
+    uint96 public englishAuctionRefreshTime;
```
Comment (LouisMeMyself) [inline contracts/JoepegAuctionHouse.sol:96]: Keep it uint256 as it will use a full storage slot either way.

You can also change `englishAuctionMinBidIncrementPct ` to be both `uint128`. If so they will be packed together saving some gas when loading them together in a function

File: contracts/JoepegAuctionHouse.sol [hunk 7]
```diff
@@ -250,35 +270,45 @@ contract JoepegAuctionHouse is
...
         uint256 _amount
-    ) public nonReentrant {
-        IERC20(_currency).safeTransferFrom(msg.sender, address(this), _amount);
-        _placeEnglishAuctionBid(_collection, _tokenId, _currency, _amount);
+    ) external nonReentrant {
+        EnglishAuction memory auction = englishAuctions[_collection][_tokenId];
+        address currency = auction.currency;
+        if (currency == address(0)) {
+            revert JoepegAuctionHouse__NoAuctionExists();
+        }
```
Comment (LouisMeMyself) [inline contracts/JoepegAuctionHouse.sol:291]: This check i redundant with L712, but I'm fine over checking things, up to you

File: contracts/JoepegAuctionHouse.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoepegAuctionHouse.sol:189]: My bad I mixed start and settle functions

File: contracts/JoepegAuctionHouse.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoepegAuctionHouse.sol:360]: I just realized that the bidder could also be a contract that can't receive AVAX directly.
Maybe the cancel could have an option to send WAVAX instead of AVAX

**Review by LouisMeMyself — CHANGES_REQUESTED** [review]

File: contracts/JoepegAuctionHouse.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoepegAuctionHouse.sol:189]: My bad I mixed start and settle functions

File: contracts/JoepegAuctionHouse.sol [hunk 8]
```diff
@@ -43,28 +45,29 @@ contract JoepegAuctionHouse is
...
+        uint96 startTime;
         address currency;
+        uint96 endTime;
         uint256 startPrice;
         uint256 endPrice;
-        uint256 startTime;
-        uint256 endTime;
-        uint256 dropInterval;
+        uint32 dropInterval;
     }
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol]: ah I see! I'm going to use 3 `uint256` then — mainly b/c `JoepegExchange` uses `uint256` for prices so I want to be consistent here

File: contracts/JoepegAuctionHouse.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:368]: hm I see — that's probably why `JoepegExchange` also just sends out WAVAX instead of AVAX. i think i'll do the same to avoid complicating the contract code

File: contracts/JoepegAuctionHouse.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol]: makes sense, thanks for the thorough explanation! i've changed them to `IERC721` type in the params

File: contracts/JoepegAuctionHouse.sol (hunk 6 above)
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:96]: I think its better to keep this as `uint96` b/c it helps me to avoid having to cast from `uint256` -> `uint96` in `_placeEnglishAuctionBid` on line 729 where I increment `auction.endTime` (uint96) by `englishAuctionRefreshTime`

**Comment by 0xnivek** [comment]
Update:
- Changed contract to send out WAVAX instead of AVAX to avoid edge cases where the receiver isn't an EOA and can't receive AVAX
- Added tests
- Added deploy script

**Review by cryptofish7 — APPROVED** [review]

**Review by simsk87 — APPROVED** [review]

File: deploy/JoepegAuctionHouse.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            args: initArgs,
+          },
+        },
+      },
+      log: true,
+      deterministicDeployment: false,
+    });
+  });
+
+  await verify(proxyContract.implementation, constructorArgs);
```
Comment (Mathieu-Be) [inline deploy/JoepegAuctionHouse.js:71]: You should add a check here that you're not on Hardhat because iirc it crashes if you're deploying on your local network

**Review by Mathieu-Be — APPROVED** [review]
LGTM

File: contracts/JoepegAuctionHouse.sol [hunk 10]
```diff
@@ -0,0 +1,1154 @@
...
+            _minPercentageToAsk * _amount
+        ) {
+            revert JoepegAuctionHouse__FeesHigherThanExpected();
+        }
+
+        // 3. Transfer final amount (post-fees) to seller
+        {
+            IERC20(WAVAX).safeTransfer(_to, finalSellerAmount);
+        }
+    }
```
Comment (LouisMeMyself) [inline contracts/JoepegAuctionHouse.sol:1169]: These 2 functions are now identical if you use WAVAX as the currency in the `_transferFeesAndFunds` function. The WAVAX one should be removed to simplify the contract

**Review by LouisMeMyself — APPROVED** [review]
LGTM

File: contracts/JoepegAuctionHouse.sol [hunk 11]
```diff
@@ -0,0 +1,1154 @@
...
+    address public protocolFeeRecipient;
+
+    /// @notice Stores Dutch Auction data for NFTs
+    /// @dev (collection address => token id => dutch auction)
+    mapping(address => mapping(uint256 => DutchAuction)) public dutchAuctions;
+
+    /// @notice Stores English Auction data for NFTs
+    /// @dev (collection address => token id => english auction)
+    mapping(address => mapping(uint256 => EnglishAuction))
+        public englishAuctions;
```
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/JoepegAuctionHouse.sol:98]: (This is not very important, but it's best code practice)

File: contracts/JoepegAuctionHouse.sol [hunk 12]
```diff
@@ -0,0 +1,1154 @@
...
+        uint256 _startPrice,
+        uint256 _endPrice,
+        uint256 _minPercentageToAsk
+    )
+        external
+        isSupportedCurrency(_currency)
+        isValidMinPercentageToAsk(_minPercentageToAsk)
+        nonReentrant
+    {
+        if (_duration == 0 || _duration < _dropInterval) {
```
[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:548]: nope! if `_dropInterval == 0`, that means there is no price drop over time in which case it prob doesnt make sense for it to be a dutch auction

[the remark this answers is not in the document]
Comment (simsk87) [reply contracts/JoepegAuctionHouse.sol:556]: do we check this anywhere then?

[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:556]: o good catch will add

[the remark this answers is not in the document]
Comment (0xnivek) [reply contracts/JoepegAuctionHouse.sol:557]: done!

**Review by LouisMeMyself — APPROVED** [review]
LGTM
